### PR TITLE
fix: 14 small triaged issues (OIDC/security hardening, loaders, logging, flaky test)

### LINKED ISSUE #121: Conflict Between `unique=True` and `nullable=True` on `external_user_id` Field Prevents User Creation
### Origin of the Error
When creating 2 users without supplying the external_user_id (so it's supposed to be None), it leads to error. Here is the definition of the field external_user_id.

https://github.com/linagora/openrag/blob/4716004453435a30c44d8577a209f22342b69e4b/openrag/components/indexer/vectordb/utils.py#L102

Although PostgreSQL allows multiple `NULL` values in a `UNIQUE` column, new users are being created with an empty string (`''`) instead of `None`.
Since PostgreSQL treats `''` as a valid (non-null) value, the `UNIQUE` constraint is enforced, leading to:

```bash
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "ix_users_external_user_id"
DETAIL: Key (external_user_id)=() already exists.
```

**Cause:**
FastAPI’s Swagger UI sends an empty string (`''`) when the field is left blank.
Only when “Send empty value” is **unchecked** does it send `None`, allowing user creation.

**Reference:**
[[openrag/routers/users.py#L28-L43](https://github.com/linagora/openrag/blob/4716004453435a30c44d8577a209f22342b69e4b/openrag/routers/users.py#L28-L43)](https://github.com/linagora/openrag/blob/4716004453435a30c44d8577a209f22342b69e4b/openrag/routers/users.py#L28-L43)

---

#### Simple solution
Ensure that empty strings are converted to `None` before database insertion:

```python
@router.post("/")
async def create_user(
    display_name: Optional[str] = Form(None),
    external_user_id: Optional[str] = Form(None),
    is_admin: bool = Form(False),
    vectordb=Depends(get_vectordb),
    admin_user=Depends(require_admin),
):
    """
    Create a new user and generate a token.
    """
    user = await vectordb.create_user.remote(
        display_name=display_name,
        external_user_id=external_user_id or None,
        is_admin=is_admin,
    )
    logger.info("Created new user", user_id=user["id"])
    return JSONResponse(status_code=status.HTTP_201_CREATED, content=user)
```

> [!NOTE]  
> This issue is related to this commit when we replaced Query to Form: https://github.com/linagora/openrag/commit/dcfd643ba1cd36a7efd4da8340c95e697523714d

### LINKED ISSUE #360: OIDC next URL accepts backslash hosts and CRLF injection
`_sanitize_next_url` in `openrag/routers/auth.py` around line 153 returns `next_url` unchanged whenever it begins with `/` and does not begin with `//`. A value like `/\evil.com/path` (which Chrome and Edge parse as a host-relative URL whose host is `evil.com`) and any value containing `%0d%0a` both pass this check, then flow into the `Location` header on the OIDC redirect. That enables open-redirect phishing and HTTP response splitting against authenticated sessions.

Reject any `next_url` whose second character is `/` or `\`, strip CR and LF before returning, and confirm with `urllib.parse.urlsplit` that the parsed `netloc` and `scheme` are empty. Add unit tests covering `/\evil.com`, `//evil.com`, and CRLF payloads.

### LINKED ISSUE #361: Admin token rotates on every startup when AUTH_TOKEN is unset
`_ensure_admin_user` in `openrag/components/indexer/vectordb/utils.py` around lines 71 to 75 generates a fresh `or-` token whenever `AUTH_TOKEN` is unset and unconditionally writes it to the existing admin row. Every container restart therefore overwrites the value previously stored, including any token issued via `POST /users/{id}/regenerate_token` for user id 1, silently invalidating all admin clients and CI integrations with no audit trail.

Only generate and persist a new admin token when the row is being created or has no token set. When `AUTH_TOKEN` is provided, write it; otherwise leave the existing token untouched and emit a single warning that no static admin token is configured. Cover both branches with a startup test that asserts the token survives a re-run of bootstrap.

### LINKED ISSUE #363: Empty LLM stream surfaces StopAsyncIteration as 500
In `openrag/routers/openai.py` at lines 358 and 434, the non-streaming chat and completion handlers call `await llm_output.__anext__()` without catching `StopAsyncIteration`. When the upstream RAG generator yields zero items (LLM upstream returned an empty body, hit a content filter, or raised mid-stream and exited cleanly), the exception propagates out of the handler as an unhandled 500 with no useful detail, and clients have no way to distinguish it from a genuine server crash.

Wrap each call in `try / except StopAsyncIteration` and respond with HTTP 502 plus a structured error explaining the upstream LLM produced no chunks. Add a unit test that mocks the generator returning zero items and asserts the handler returns 502, not 500.

### LINKED ISSUE #376: Multi-page document loader keeps only the final page
`openrag/components/indexer/loaders/CustomDocLoader.py:37` builds page text inside the page loop with `s = p.page_content.strip() + f"\n[PAGE_{page_num}]\n"`. The assignment overwrites `s` on every iteration, so when `load()` returns `s` at the end of the loop only the last page survives. Every multi-page legacy `.doc` document indexed via this loader silently loses every page but the final one, with no error visible in logs.

Change the assignment to `s += p.page_content.strip() + f"\n[PAGE_{page_num}]\n"`. Add a unit test that loads a fixture with at least three pages and asserts each page's text is present in the returned content.

### LINKED ISSUE #386: OIDC userinfo claims accepted without sub binding to ID token
`openrag/routers/auth.py:309-323` reads `claims_for_mapping` from the `/userinfo` endpoint when `OIDC_CLAIM_SOURCE=userinfo` and feeds them straight into the claim-mapping write path. OIDC Core 1.0 section 5.3.2 requires the RP to verify that `userinfo.sub == id_token.sub` before trusting any userinfo claim. Without that check, a malformed userinfo response or a token-substitution attack at the userinfo endpoint can rewrite the matched user's `display_name` and `email` on every login.

Compare `claims_for_mapping["sub"]` to `bundle.claims["sub"]` immediately after fetching userinfo and abort the callback with a 401 on mismatch. Log the mismatch with the sub values masked so operators can detect a misconfigured IdP without leaking identifiers.

### LINKED ISSUE #426: Flaky API test: test_cancel_increments_total_cancelled races the async counter update
## Summary

`tests/api_tests/test_indexer.py::TestTaskCancellation::test_cancel_increments_total_cancelled` is flaky and broke the API Tests workflow on the `chore: bump version to 1.1.11` push to `main` (commit c9d53cc0, run [26233866578](https://github.com/linagora/openrag/actions/runs/26233866578)).

The tag-triggered `Build and Release` workflow itself succeeded — only this test failed.

## Failure

```
FAILED test_indexer.py::TestTaskCancellation::test_cancel_increments_total_cancelled
  assert 1 == (1 + 1)
```

## Root cause

At `tests/api_tests/test_indexer.py:766-791` the test:

1. Reads `cancelled_before = /queue/info → tasks.total_cancelled`
2. Uploads a file and `DELETE /indexer/task/{task_id}`
3. Immediately re-reads `cancelled_after`
4. Asserts `cancelled_after == cancelled_before + 1`

The `DELETE` endpoint only returns *"Cancellation signal sent"*. The `total_cancelled` counter is bumped asynchronously by `TaskStateManager` when the task actually transitions to `CANCELLED`. The test re-reads `/queue/info` before that bump lands, so the counter looks unchanged.

The previous test in the class (`test_cancel_task_sets_cancelled_state`) also cancels a task — that cancellation is what makes `cancelled_before == 1` here, confirming the counter does work, just not synchronously.

## Suggested fix

Poll `/queue/info` (or `/indexer/task/{task_id}` until `task_state == "CANCELLED"`) before asserting the counter, instead of reading immediately after the cancel call returns.

### LINKED ISSUE #477: security follow-ups: unvalidated source_file_id in copy endpoint + Ray dashboard exposed in 3 compose files
Follow-up items from the security PR sweep merged 2026-06-15 (#466–#476). Neither undoes a merged fix; both are incomplete coverage of an already-shipped fix.

## 1. Unvalidated \`source_file_id\` in copy-file endpoint (follow-up to #470)

#470 added allowlist validation (\`[A-Za-z0-9._:\-]+\`) for \`file_id\`/partition names to block Milvus filter injection, but the **copy-file** path was missed:

- \`source_file_id\` (Form field in \`copy_file_between_partitions\`, \`openrag/routers/indexer.py\`) flows unvalidated into \`indexer.copy_file\` → \`vectordb.get_file_chunks\` → \`filter_expr = f'partition == "{partition}" and file_id == "{file_id}"'\` (\`vectordb.py\`).
- An attacker able to call the copy endpoint can inject via \`source_file_id\` (e.g. \`x" or partition == "other\`), the same injection class #470 closed for path params.

**Fix:** validate \`source_file_id\` with the same \`assert_valid_file_id\`/allowlist helper before it reaches the filter. Consider also validating the \`file_ids\` list body in \`add_files_to_workspace\` (defense-in-depth) and dropping the now-dead \`FORBIDDEN_CHARS_IN_FILE_ID\` constant in \`indexer.py\`.

## 2. Ray dashboard still publicly exposed in 3 compose files (follow-up to #471)

#471 added the \`127.0.0.1:\` localhost bind for the Ray dashboard (port 8265) only in the root \`docker-compose.yaml\`. These functional compose files still map \`8265\` to all interfaces:

- \`quick_start/docker-compose.yaml\` (\`\${RAY_DASHBOARD_PORT:-8265}:8265\`)
- \`docs/assets/compose_linux_gpu.yaml\` (\`\${RAY_DASHBOARD_PORT:-8265}:8265\`)
- \`docs/assets/compose_ollama_cpu.yaml\` (\`8265:8265\`)

The Ray dashboard has no auth and exposes a job-submission RCE API, so any deployment started from these files remains exploitable.

**Fix:** apply the same \`127.0.0.1:\` host-side bind to the dashboard port mapping in all three files.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### LINKED ISSUE #481: [security] Raw user search queries logged at INFO (privacy; inconsistent with M10)
**Severity:** Medium

`openrag/routers/search.py` binds `query=search_params.text` (lines 129, 149, 239, 258, 339, 345) and emits `log.info(...)` (lines 156, 171, 266, 278, 352). The loguru formatter renders `extra`, so raw user search queries land in `logs/app.json` at the default INFO level.

The chat path was fixed under M10 to log queries only at DEBUG (`pipeline.py:284`, `map_reduce.py:156`); the search router was missed, so M10 is only half-applied.

**Impact:** user search queries (potentially PII / sensitive) persisted in plaintext logs.

**Fix:** drop `query=` from the INFO binds in `search.py` (log it at DEBUG only, matching the chat path).

### LINKED ISSUE #483: [security] Open redirect via backslash in OIDC `next` parameter
**Severity:** Low

`openrag/routers/auth.py` `_sanitize_next_url` accepts any value starting with `/` but not `//`:
```python
if next_url.startswith("/") and not next_url.startswith("//"):
    return next_url
```
A value like `/\evil.com` passes (`startswith("/")` true, `startswith("//")` false). Browsers normalize `\` → `/`, treating `/\evil.com` as the protocol-relative `//evil.com` → redirect to an attacker origin after a successful login.

**Exploit:** send a victim `GET /auth/login?next=/\evil.com`; the callback issues `302 Location: /\evil.com`.

**Fix:** reject `next_url` containing `\` (and ideally any control/whitespace char) before the relative-path branch, or normalize backslashes before the `//` check.

### LINKED ISSUE #485: [security] OIDC: missing `azp` check on multi-audience ID tokens
**Severity:** Low (defense-in-depth)

`openrag/components/auth/oidc_client.py` accepts an ID token whose `aud` is a list containing the client_id, but never checks `azp`. Per OIDC Core §3.1.3.7, when multiple audiences are present the `azp` (authorized party) claim must be present and equal to the client_id.

**Impact:** narrow — requires an IdP configured to mint multi-audience tokens; a token legitimately issued to another relying party (but listing openrag in `aud`) could be replayed at the callback.

**Fix:** when `aud` is a list with more than one entry, require `azp == client_id`. Apply to both `_verify_id_token` and `verify_logout_token`.

### LINKED ISSUE #487: [security] Control-token neutralizer (H8) narrower than its output parser
**Severity:** Low (defense-in-depth)

`neutralize_prompt_control_tokens` (`openrag/components/indexer/utils/text_sanitizer.py`) is narrower than the output parser it protects, `_SOURCES_NUMS_RE` (`openrag/components/utils.py:180`). For example `[Sources: 1, 2]` is neutralized to `(Sources: 1, 2]`, which the parser still matches (no leading-`[` requirement / no anchor). A tag with leading prose on a line is not touched at all.

**Impact:** low — the strong `[Source N]` block boundary marker *is* reliably defanged, so fake source *boundaries* in context are blocked. The residual gap is citation-attribution forgery/suppression if the LLM echoes a document-supplied `Sources:` tag (UI source list manipulation only, no data exposure).

**Fix:** tighten the neutralizer regexes to match the parser, or (better) anchor the *output* parser to require the application's own `[Source]` framing so document-echoed tags can't match.

### LINKED ISSUE #488: [security] Ansible deploy hardening: host key checking disabled, world-readable .env, unpinned installer
**Severity:** Low

`ansible/`:

- **SSH host key verification disabled** — `ansible.cfg:3,10`: `host_key_checking = False` plus `UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no`. The deploy copies a secrets-bearing `.env` over an unverified SSH session (MITM exposure).
- **`.env` written world-readable (0644)** — `playbooks/openrag.yml:101-107` writes the secrets file (`POSTGRES_PASSWORD`, MinIO keys, `AUTH_TOKEN`, OIDC secret/encryption key, HF token) as mode `0644`; any local user on the host can read it. Use `0600`.
- **Unpinned `curl ... | sh`** — `playbooks/openrag.yml:127` pipes a remote uv installer script into a shell with no version pin/checksum.

**Fix:** enable host key checking (pre-seed known_hosts), set the `.env` file mode to `0600`, and pin/checksum the uv installer.

### PR DESCRIPTION
**Pull request by EnjoyBacon7**
## What this does

Fixes a batch of small, self-contained issues found during the recent open-issue triage. Each change is under 20 lines. Grouped by area below.

## Auth and OIDC

- Reject backslashes and CR/LF in the OIDC `next` URL, so `/\evil.com` and header-injection payloads can't slip past the open-redirect guard. (#360, #483)
- Require the `/userinfo` `sub` to match the verified ID-token `sub` before applying claim mapping. (#386)
- Enforce `azp == client_id` on multi-audience ID and logout tokens. (#485)
- Stop rotating the admin token on every startup when `AUTH_TOKEN` is unset; keep the stored one. (#361)
- Revoke a user's active OIDC sessions when their API token is regenerated, via a new reusable `revoke_oidc_sessions_by_user_id`. (partial #486)

## Indexing and loaders

- Accumulate every page in CustomDocLoader instead of keeping only the last one. (#376)
- Coerce an empty `external_user_id` to NULL so it can't collide on the unique index. (#121)

## API and deploy

- Validate `source_file_id` in the copy endpoint, and bind the Ray dashboard to localhost in the last compose file that still exposed it. (#477)
- Widen the prompt-injection neutralizer so a forged sources tag can't survive it. (#487)
- Return 502 instead of an unhandled 500 when the upstream LLM stream is empty. (#363)
- Stop logging the raw search query at INFO; log its length instead. (#481)
- Harden the Ansible deploy: host-key checking on, `.env` at 0600, pinned uv installer. (#488)

## Tests

- De-flake `test_cancel_increments_total_cancelled` by polling `/queue/info` instead of reading the counter right after the delete. (#426)

## Testing

- `uv run pytest openrag/` → 386 passed, 5 skipped
- `uv run ruff check openrag/` and `ruff format --check openrag/` → clean

Fixes #360, fixes #483, fixes #386, fixes #485, fixes #361, fixes #376, fixes #121, fixes #477, fixes #487, fixes #363, fixes #481, fixes #488, fixes #426.

### CODE REVIEW

File: ansible/ansible.cfg [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
-host_key_checking = False
+host_key_checking = True
 retry_files_enabled = False
 stdout_callback = default
 gathering = smart
 fact_caching = memory
 
 [ssh_connection]
-ssh_args = -o ControlMaster=auto -o ControlPersist=60s -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no
+ssh_args = -o ControlMaster=auto -o ControlPersist=60s -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=accept-new
```
[the remark this answers is not in the document]
Comment (EnjoyBacon7) [reply ansible/ansible.cfg]: Good catch — removed `UserKnownHostsFile=/dev/null` so `StrictHostKeyChecking=accept-new` can actually persist and validate host keys across sessions. Folded into the Ansible hardening commit.

File: ansible/playbooks/openrag.yml [hunk 2]
```diff
@@ -124,7 +124,7 @@
     - name: Setup Python environment
       block:
         - name: Install uv (Python package manager)
-          shell: curl -LsSf https://astral.sh/uv/install.sh | sh
+          shell: curl -LsSf https://astral.sh/uv/0.5.11/install.sh | sh
           args:
             creates: "/home/{{ project_user }}/.cargo/bin/uv"
```
[the remark this answers is not in the document]
Comment (EnjoyBacon7) [reply ansible/playbooks/openrag.yml]: Done — the uv install now runs under `/bin/bash` with `set -euo pipefail`, so a failed `curl` is no longer masked by the pipe. Folded into the Ansible hardening commit.

File: tests/api_tests/test_indexer.py [hunk 3]
```diff
@@ -784,9 +784,16 @@ def test_cancel_increments_total_cancelled(self, api_client, created_partition,
...
+        start = time.time()
+        while time.time() - start < TASK_TIMEOUT:
+            info_after = api_client.get("/queue/info")
+            assert info_after.status_code == 200
+            cancelled_after = info_after.json()["tasks"]["total_cancelled"]
+            if cancelled_after >= cancelled_before + 1:
+                break
+            time.sleep(0.5)
 
         assert cancelled_after == cancelled_before + 1
```
[the remark this answers is not in the document]
Comment (EnjoyBacon7) [reply tests/api_tests/test_indexer.py:799]: Done — relaxed the final assertion to `>=` to match the polling condition, since `total_cancelled` is a shared async counter. Folded into the de-flake commit.

**Comment by EnjoyBacon7** [comment]
CodeRabbit review addressed:

- **3 inline comments** fixed and folded into their commits: dropped `UserKnownHostsFile=/dev/null` so `accept-new` actually validates host keys; ran the uv install under `/bin/bash` with `set -euo pipefail`; relaxed the cancel-counter assertion to `>=` (shared async counter). See thread replies.
- **Possibly-related PRs #400 / #405**: both merged into `refactor/hexagonal`, not `main`. `main` still carries both bugs (CustomDocLoader still overwrites pages; no `azp` check), so the changes here are the equivalent fixes for `main` and will reconcile trivially when the refactor lands.
- **Docstring Coverage warning (42.86%)**: left as-is. The shortfall is in pre-existing functions in the touched files; adding docstrings there would be outside the scope of these fixes (which the Out-of-Scope check otherwise passed).
